app/services/authentication/strategy/infer_service.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module Authentication
+  module Strategy
+    class InferService < ThreeScale::Patterns::Service
+
+      def initialize(params, account, provider)
+        @params = params
+        @account = account
+        @provider = provider
+      end
+
+      def call
+        type = infer_type
+
+        return build(:token) if impersonating?(type)
+        return build(:provider_oauth2) if sso_enforced?
+
+        build(type)
+      end
+
+      private
+
+      STRATEGIES = %i[oauth2_base internal token].freeze
+
+      def infer_type
+        given_param_keys = @params.to_h.symbolize_keys.keys
+
+        # Take the first strategy in STRATEGIES which expected_params are received in the request
+        # Ignore :request and :redirect_url params since they are not used for authentication
+        STRATEGIES.find { (given_param_keys - %i[request redirect_url] - strategy_class(_1).expected_params).empty? }
+      end
```

Comment:
My gut feeling is that something is not alright with inferring the auth strategy from user input. Isn't there a way to know which strategy is supposed to work with a particular provider/dev portal?

Reply:
It is not a clear cut here. My feeling is that in case we go over the strategies in a certain order, that will simplify the logic here and as a lesser/negligible concern, that would make sure strategies with conflicting parameters would not break things.

But I don't have a hard opinion, just perhaps a small preference. So you can keep as is.

Reply:
Yeah, I've been trying the second approach but abandoned it because of how the controllers use the strategy. They first get the strategy, then call `authenticate` on it. If we want to know which strategy succeeds and return that one, we must call `authenticate` from `InferService`, and then return the `user` and the `strategy` to the controller. It can be done but requires more work and I don't think we'll get much benefit from it.

I made some changes here anyway. I fixed a bug: when no parameters were received, the inferrer was returning always `oauth2_base` as the strategy. Now it returns `nil` in this situation. I had to add some additional code to gracefully manage the exception.

https://github.com/3scale/porta/pull/4064/commits/640baf6bef95ac59bc7050fb65e2bb1c279ad4f8